### DISCUSSION TITLE
Should we use DeepCopy() to update Spec?

### DISCUSSION THREAD

**Discussion by hatrena** [kubernetes-sigs/kubebuilder#2804] [Support] [answered]
Hi, 

I am new to Kubebuilder, so the question is "Should I use DeepCopy() to update the object's Spec?". if so, why?

sample code:

```
objToUpdate := obj.DeepCopy()
objToUpdate.Spec.Foo = "bar"

err := r.Client.Patch(ctx, objToUpdate, client.MergeFrom(obj))
if err != nil {
return ctrl.Result{}, fmt.Errorf("error on patch obj %w", err)
}
```

**Comment by FillZpp** [comment, marked as answer]
This is not necessary. Only note that both Update and Patch methods will overwrite the new object returned from APIServer into the given object, which means the `objToUpdate` will be changed to the latest object after patched successfully.

So, you should DeepCopy it only if you don't want the original object to be modified.

**Reply by hatrena** [reply]
Got it. Thank you so much.

**Comment by hatrena** [comment]
So everything is clear. I have tested it and watched `resourceVersion` of the object. With DeepCopy() the original object (in the code) doesn't get modified (but it has updated in `etcd` as `resourceVersion` increased). 

So in what cases I don't need the original to get updated in my code? Do you have any example scenarios? 

I appreciate your help

**Comment by FillZpp** [comment]
> So in what cases I don't need the original to get updated in my code? Do you have any example scenarios?

If you ever read the KCM code, you may find those built-in controllers always deep copy objects before modify them. That's because those objects are references that listed from Informer directly, which means modify them will lead to the objects in Informer cache been changed unexpectedly.

But controller-runtime, on the other hand, will deep copy objects  before returning them to List method by default, which means the objects you listed are already deep copied. So you can directly pass them into Update or Patch, unless you have used the `UnsafeDisableDeepCopyByObject` option to disable the default DeepCopy for List operation.

https://github.com/kubernetes-sigs/controller-runtime/blob/15154aaa67679df320008ed45534f83ff3d6922d/pkg/cache/cache.go#L126-L130

**Comment by camilamacedo86** [comment]
Closing since it is answered.
